protocols/gossipsub/src/behaviour.rs

```diff
@@ -947,6 +943,17 @@ where
                     topic_hash: topic_hash.clone(),
                 },
             );
+
+            // If the peer did not previously exist in any mesh, inform the handler
+            if self.peer_added_to_mesh(&peer_id, topic_hash) {
```

Comment:
These _check if peer was added and if so inform the handler_ blocks are quite pervasive. I am not familiar enough with the codebase to tell whether they are inherent, or whether there is room for improvement. Would e.g. extracting the _mesh tracking_ logic to a separate component be an option? Each time one instructs the separate component to add a peer to a mesh, it returns (with `#[must-use]`) whether the corresponding handler needs to be informed.

Reply:
Yes. I agree. 

I found it difficult to find a good pattern for this. Originally these were mutable functions which handled sending the events, but its difficult to inject them somewhat generically in the places they need to be with a mutable reference to self (even if I don't mutate the fields being borrowed). I resorted to immutable reference to self as a work-around. 

Ideally a new structure of the mesh which handles this internally might be a nicer approach, but there are parts of the code that take mutable references to the peers in each mesh and mutate them. I'll have play try and refactor it a little